### DISCUSSION TITLE
Custom 404 for static file server

### DISCUSSION THREAD

**Discussion by Umenokin** [tokio-rs/axum#331] [Q&A] [answered]
I am following the "static-file-server" example but do not understand how can I customize 404 error response. For example, I would want to return a custom HTML page. From the ServeDir source code below, it does not look like you can customize ServeDir itself:

```
fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
        match &mut self.inner {
            Inner::Valid(open_file_future) => {
                let (file, mime) = match ready!(Pin::new(open_file_future).poll(cx)) {
                    Ok(Output::File(file, mime)) => (file, mime),
               ...
                Poll::Ready(Ok(res))
            }
            Inner::Invalid => {
                let res = Response::builder()
                    .status(StatusCode::NOT_FOUND)
                    .body(empty_body())
                    .unwrap();

                Poll::Ready(Ok(res))
            }
        }
    }
```

So, I am trying to intercept the response somehow. My current attempt is to use custom service with layer but could not make it to work:
```
Router::new()
        .layer(NotFoundLayer::new())
        .nest("/", service::get(file_service).handle_error(|error: std::io::Error| {
            println!("Error: {}", error);
            Ok::<_, Infallible>((
                StatusCode::INTERNAL_SERVER_ERROR,
                format!("Unhandled internal error: {}", error),
            ))
        }))
        //.or(handler_404.into_service())
        .boxed()
```
In this example, nothing would call NotFound's `call` if the file was not found. 

Is there any other option I could use?

Thanks!

**Comment by davidpdrsn** [comment, marked as answer]
If you want a customize 404 response that only applies to `ServerDir` then you have to route to a service that internally calls `ServeDir` and changes the response as necessary. There is an example of doing that here https://github.com/tokio-rs/axum/issues/87#issuecomment-891326336

**Reply by davidpdrsn** [reply]
`handle_error` is only for handling errors. A 404 response is not considered an error. Only [`tower::Service::Error`s](https://docs.rs/tower/0.4.8/tower/trait.Service.html) are considered errors.

**Reply by Umenokin** [reply]
Thanks again, @davidpdrsn , You save my day, and my motivation. It is really hard digging through endless not-bound errors!
